### PR TITLE
Interaction model

### LINKED ISSUE #4: Parse whatsaap notifications

### LINKED ISSUE #5: Manage db for when last contacted contacts

### PR DESCRIPTION
**Pull request by roy-shchory**

### CODE REVIEW

File: app/src/main/java/com/cortado/mafkir/di/ViewModelFactoryModule.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import androidx.lifecycle.ViewModelProvider
+import com.cortado.mafkir.util.ViewModelProviderFactory
+import dagger.Binds
+import dagger.Module
+
+
+@Module
+abstract class ViewModelFactoryModule {
+
+    // Method #2
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/di/ViewModelFactoryModule.kt]: ?

File: app/src/main/java/com/cortado/mafkir/di/ViewModelFactoryModule.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    @Binds
+    abstract fun bindViewModelFactory(viewModelProvideFactory: ViewModelProviderFactory): ViewModelProvider.Factory
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/di/ViewModelFactoryModule.kt]: `viewModelProvideFactory`: did you mean `viewModelProviderFactory`?

File: app/src/main/java/com/cortado/mafkir/di/ViewModelModule.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import com.cortado.mafkir.model.MafkirContactViewModel
+import dagger.Binds
+import dagger.Module
+import dagger.multibindings.IntoMap
+
+
+@Module
+abstract class ViewModelModule {
+
+    // Method #1
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/di/ViewModelModule.kt]: ?

File: app/src/main/java/com/cortado/mafkir/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import com.cortado.mafkir.permissions.MafkirPermissionsValidator
+import com.cortado.mafkir.util.ViewModelProviderFactory
+import dagger.android.support.DaggerAppCompatActivity
+import javax.inject.Inject
+import javax.inject.Named
+
+
+class MainActivity : DaggerAppCompatActivity() {
+    @field:[Inject Named("User")]
+    lateinit var mafkirNotifier: MafkirNotifier
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/MainActivity.kt]: unused

File: app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    override fun onCreate() {
+        AndroidInjection.inject(this)
+        super.onCreate()
+    }
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        super.onStartCommand(intent, flags, startId)
+        val notification = mafkirNotifier.build(
+            "Mafkir",
+            "mafkir has got your back, were here to help you not forget",
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt]: `Mafkir... we are...`

File: app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            true
+        )
+        startForeground(1337, notification)
+        return START_STICKY
+    }
+
+    override fun onNotificationRemoved(sbn: StatusBarNotification) {
+        //
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt]: you meant `do nothing`?

File: app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    override fun onNotificationPosted(sbn: StatusBarNotification) {
+        // TODO: extract to subscribers
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt:42]: if you meant `rewrite and separate to like 3 different classes` then that's a good todo, and ignore this comment

File: app/src/main/java/com/cortado/mafkir/MainActivity.kt [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        startService(Intent(this, InteractionsService::class.java))
+
+        setContentView(R.layout.activity_main)
+
+        mafkirContactViewModel = ViewModelProvider(
+            this,
+            viewModelProviderFactory
+        ).get(MafkirContactViewModel::class.java)
+
+        mafkirContactViewModel.getAll().observe(this, Observer { lisOfNotes ->
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/MainActivity.kt]: `lisOfNotes` -> contactsList?

File: app/src/main/java/com/cortado/mafkir/MainActivity.kt [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+            lisOfNotes?.let {
+                lisOfNotes.iterator().forEach {
+                    Log.i("Mafkir", "contact:$it")
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/MainActivity.kt]: can probably be deleted if tested 7 times

File: app/src/main/java/com/cortado/mafkir/MainActivity.kt [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                }
+            }
+        })
+
+        val button = findViewById<Button>(R.id.testNotification)
+        button.setOnClickListener {
+            //
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/MainActivity.kt]: huh

File: app/src/main/java/com/cortado/mafkir/util/ViewModelProviderFactory.kt [hunk 11]
```diff
@@ -0,0 +1,26 @@
+package com.cortado.mafkir.util
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/util/ViewModelProviderFactory.kt]: repackage

File: app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+        if (sbn.packageName == "com.whatsapp") {
+            sbn.notification.extras.getCharSequence(
+                Notification.EXTRA_TITLE
+            )?.toString()?.apply {
+                mafkirContactRepository.updateLastInteraction(this)
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt:47]: do u really wanna do this without checking if the contact was chosen by the user as contact he wants to keep in contact with?

File: app/src/main/java/com/cortado/mafkir/di/ViewModelFactoryModule.kt [hunk 13]
```diff
@@ -0,0 +1,15 @@
+package com.cortado.mafkir.di
+
+import androidx.lifecycle.ViewModelProvider
+import com.cortado.mafkir.util.ViewModelProviderFactory
+import dagger.Binds
+import dagger.Module
+
+
+@Module
+abstract class ViewModelFactoryModule {
```
Comment (IdoAsraff) [inline app/src/main/java/com/cortado/mafkir/di/ViewModelFactoryModule.kt]: absolutely unused if im not mistaken

File: app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt (hunk 12 above)
Comment (roy-shchory) [reply app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt:47]: if it wasent chosen by the user nothing happens

File: app/src/main/java/com/cortado/mafkir/di/ViewModelFactoryModule.kt (hunk 1 above)
Comment (roy-shchory) [reply app/src/main/java/com/cortado/mafkir/di/ViewModelFactoryModule.kt]: d

File: app/src/main/java/com/cortado/mafkir/di/ViewModelFactoryModule.kt (hunk 2 above)
Comment (roy-shchory) [reply app/src/main/java/com/cortado/mafkir/di/ViewModelFactoryModule.kt]: d

File: app/src/main/java/com/cortado/mafkir/MainActivity.kt (hunk 4 above)
Comment (roy-shchory) [reply app/src/main/java/com/cortado/mafkir/MainActivity.kt]: d

File: app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt (hunk 5 above)
Comment (roy-shchory) [reply app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt]: d

File: app/src/main/java/com/cortado/mafkir/MainActivity.kt (hunk 10 above)
Comment (roy-shchory) [reply app/src/main/java/com/cortado/mafkir/MainActivity.kt]: ignore will be removed when i add the gui

File: app/src/main/java/com/cortado/mafkir/MainActivity.kt (hunk 9 above)
Comment (roy-shchory) [reply app/src/main/java/com/cortado/mafkir/MainActivity.kt]: removed

File: app/src/main/java/com/cortado/mafkir/MainActivity.kt (hunk 8 above)
Comment (roy-shchory) [reply app/src/main/java/com/cortado/mafkir/MainActivity.kt]: removed

File: app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt (hunk 7 above)
Comment (roy-shchory) [reply app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt:42]: wanker

File: app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt (hunk 6 above)
Comment (roy-shchory) [reply app/src/main/java/com/cortado/mafkir/notifications/InteractionsService.kt]: d

File: app/src/main/java/com/cortado/mafkir/util/ViewModelProviderFactory.kt (hunk 11 above)
Comment (roy-shchory) [reply app/src/main/java/com/cortado/mafkir/util/ViewModelProviderFactory.kt]: d

File: app/src/main/java/com/cortado/mafkir/di/ViewModelFactoryModule.kt (hunk 13 above)
Comment (roy-shchory) [reply app/src/main/java/com/cortado/mafkir/di/ViewModelFactoryModule.kt]: lols

**Pull request merged**